### PR TITLE
Refactor code for reusability

### PR DESCRIPTION
**Pull request by vishnuravi**
# Refactor for reusability

## :recycle: Current situation & Problem
Currently, all of the OpenAI querying logic is run inside `MainInputView`. We need to separate the UI and the logic in such a way that it can be easily updated, extended, and reused.

## :bulb: Proposed solution
- Creates a `MessageManager` passed as an EnvironmentObject that manages the state of queries and messages.
- Creates an `OpenAIManager` that manages interactions with OpenAI's API.
- Creates a `PromptGenerator` to assemble prompts in a modular fashion.
- Refactors health data fetching code to use Swift concurrency.
- Fixes error in sleep data calculation.

Thank you @PSchmiedmayer for your assistance with the above!

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [X] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: HealthGPT/HealthGPT/HealthDataFetcher.swift [hunk 1]
```diff
@@ -64,120 +67,151 @@ class HealthDataFetcher {
...
+                    continuation.resume(returning: dailyData)
+                } else {
+                    continuation.resume(throwing: HealthDataFetcherError.resultsNotFound)
+                }
             }
-        }
 
-        healthStore.execute(query)
+            healthStore.execute(query)
+        }
```
Comment (PSchmiedmayer) [inline HealthGPT/HealthGPT/HealthDataFetcher.swift:95]: While this is a valid and functionally correct way to implement it. Apple actually provides even more convenient async overloads for the functions that you can use here. You can find a good example using `HKStatisticsCollectionQueryDescriptor` here: https://developer.apple.com/documentation/healthkit/hkstatisticscollectionquerydescriptor

File: HealthGPT/HealthGPT/HealthDataFetcher.swift [hunk 2]
```diff
@@ -64,120 +67,151 @@ class HealthDataFetcher {
...
     }
 
+    /// Fetches the user's health data for the specified category type identifier for the last two weeks.
+    ///
+    /// - Parameter identifier: The `HKCategoryTypeIdentifier` representing the type of health data to fetch.
+    /// - Returns: An array of `Double` values representing the daily health data for the specified identifier.
+    /// - Throws: `HealthDataFetcherError` if the data cannot be fetched.
```
Comment (PSchmiedmayer) [inline HealthGPT/HealthGPT/HealthDataFetcher.swift:102]: Thank you for the code comments, great to see here 🚀

File: HealthGPT/HealthGPT/HealthDataFetcher.swift [hunk 3]
```diff
@@ -64,120 +67,151 @@ class HealthDataFetcher {
...
-                for sample in samples {
-                    let startOfSampleDay = Calendar.current.startOfDay(for: sample.startDate)
-                    let distance = Int(Date().timeIntervalSince(startOfSampleDay) / 86400)
-                    guard let minutes = Calendar.current.dateComponents(
-                        [.minute],
-                        from: sample.startDate,
-                        to: sample.endDate
-                    ).minute else {
-                        return
+        return try await withCheckedThrowingContinuation { continuation in
```
Comment (PSchmiedmayer) [inline HealthGPT/HealthGPT/HealthDataFetcher.swift:112]: Similar to the implementation above, using `HKSampleQueryDescriptor` https://developer.apple.com/documentation/healthkit/hksamplequerydescriptor might be a good point here.

File: HealthGPT/HealthGPT/HealthGPTView.swift [hunk 4]
```diff
@@ -4,27 +4,64 @@
...
+import CardinalKitFHIR
+import CardinalKitSecureStorage
 import HealthKit
 import OpenAI
 import SwiftUI
 
 
 struct HealthGPTView: View {
-    @State private var userMessage: String = ""
+    @AppStorage(StorageKeys.openAIModel) var openAIModel = Model.gpt3_5Turbo
```
Comment (PSchmiedmayer) [inline HealthGPT/HealthGPT/HealthGPTView.swift:15]: Nice, good use of App Storage in the app here.

File: HealthGPT/HealthGPT/HealthGPTView.swift [hunk 5]
```diff
@@ -4,27 +4,64 @@
...
+                    alertText = "Could not find a valid API key."
+                    self.showAlert.toggle()
+                    return
+                }
+
+                // Assign the api key to the message handler
+                messageManager.updateAPIToken(apiKey)
+
+                // Set the OpenAI model in the message handler
+                messageManager.updateOpenAIModel(openAIModel)
```
Comment (PSchmiedmayer) [inline HealthGPT/HealthGPT/HealthGPTView.swift:63]: From a code styling perspective and to improve the readability of views, I would suggest to transform this in a private function of the `HealthGPTView` view that is then passed to the `onAppear` view modifier.

File: HealthGPT/HealthGPT/MessageInputView.swift [hunk 6]
```diff
@@ -35,149 +25,13 @@ struct MessageInputView: View {
                 .background(Color(.systemGray6))
                 .cornerRadius(10)
                 .lineLimit(1...5)
-                .disabled(isQuerying == true)
+                .disabled(messageManager.isQuerying)
 
             Button(action: {
-                isQuerying = true
```
Comment (PSchmiedmayer) [inline HealthGPT/HealthGPT/MessageInputView.swift:41]: Nice, thank you for cleaning up this long closure here!
We can therefore probably reenable the SwiftLint warnings for long closures in this project.

File: HealthGPT/HealthGPT/MessageManager.swift [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+        self.messages.append(newMessage)
+
+        do {
+            let healthData = try await healthDataFetcher.fetchAndProcessHealthData()
+
+            let generator = PromptGenerator(with: healthData)
+            let mainPrompt = generator.buildMainPrompt()
+
+            Task {
```
Comment (PSchmiedmayer) [inline HealthGPT/HealthGPT/MessageManager.swift:60]: A more meta-level comment here: I would either put all the work in a single task to perform everything within the task context, making this function itself not async (basically just kicking off a task), or remove the task and also await the response from OpenAI.

 If we do the second option, we could, e.g., enable UI elements to show the progress more easily as they can change their state before and after the async call to this function.

File: HealthGPT/HealthGPT/OpenAIManager.swift [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+import Foundation
+import OpenAI
+
+enum OpenAIAPIError: Error {
+    case noAPIToken
+}
+
+/// `OpenAIManager` is a class responsible for managing the interactions with the OpenAI API.
+class OpenAIManager {
```
Comment (PSchmiedmayer) [inline HealthGPT/HealthGPT/OpenAIManager.swift:15]: A nice abstraction, very cool!

File: HealthGPT/HealthGPT/PromptGenerator.swift [hunk 9]
```diff
@@ -0,0 +1,53 @@
+//
+// This source file is part of the Stanford HealthGPT project
+//
+// SPDX-FileCopyrightText: 2023 Stanford University & Project Contributors
+//
+
+import Foundation
+
+
+class PromptGenerator {
```
Comment (PSchmiedmayer) [inline HealthGPT/HealthGPT/PromptGenerator.swift:10]: Nice use of the builder pattern.

**Review by PSchmiedmayer — CHANGES_REQUESTED** [review]
Thank you for the refactoring @vishnuravi, a great improvement to the previous structure!

I had a few comments here and there, especially in the HealthKit code. Apart from this, it looks great, looking forward to merging the PR soon 🚀

File: HealthGPT/HealthGPT/OpenAIManager.swift (hunk 8 above)
Comment (vishnuravi) [reply HealthGPT/HealthGPT/OpenAIManager.swift:15]: Thank you!

File: HealthGPT/HealthGPT/MessageManager.swift (hunk 7 above)
Comment (vishnuravi) [reply HealthGPT/HealthGPT/MessageManager.swift]: Thank you for the feedback! I have removed the Task as you suggested: [e1c0658](https://github.com/StanfordBDHG/HealthGPT/pull/23/commits/e1c06584c1c26505feb21e61497a0293f08de804)

File: HealthGPT/HealthGPT/MessageInputView.swift (hunk 6 above)
Comment (vishnuravi) [reply HealthGPT/HealthGPT/MessageInputView.swift:41]: I have removed the comment disabling `closure_body_length`. Thanks!

File: HealthGPT/HealthGPT/HealthGPTView.swift (hunk 5 above)
Comment (vishnuravi) [reply HealthGPT/HealthGPT/HealthGPTView.swift]: I agree, this is much cleaner. I've made the change here: [c0b236c](https://github.com/StanfordBDHG/HealthGPT/pull/23/commits/c0b236cb7c25d0a05da857d97cf43d523c2d61cb). Thank you for the suggestion!

File: HealthGPT/HealthGPT/HealthDataFetcher.swift [hunk 10]
```diff
@@ -6,54 +6,57 @@
...
-            heartRate,
-            sleepAnalysis
-        ]
+            let types: Set = [
+                stepCount,
+                appleExerciseTime,
+                bodyMass,
+                heartRate,
+                sleepAnalysis
+            ]
```
[the remark this answers is not in the document]
Comment (vishnuravi) [reply HealthGPT/HealthGPT/HealthDataFetcher.swift]: Thank you for the suggestion! Definitely makes the code cleaner!

File: HealthGPT/HealthGPT/HealthDataFetcher.swift (hunk 1 above)
Comment (vishnuravi) [reply HealthGPT/HealthGPT/HealthDataFetcher.swift]: Thank you @PSchmiedmayer, I've implemented this!

File: HealthGPT/HealthGPT/HealthDataFetcher.swift (hunk 3 above)
Comment (vishnuravi) [reply HealthGPT/HealthGPT/HealthDataFetcher.swift:97]: Thanks @PSchmiedmayer, I'm having some issues getting this function to work. If you could take another look for me, would really appreciate it.

Comment (PSchmiedmayer) [reply HealthGPT/HealthGPT/HealthDataFetcher.swift:97]: @vishnuravi I will take a look at this in more detail and will get back to you 👍

**Review by PSchmiedmayer — APPROVED** [review]
@vishnuravi I fixed the sleep calculation by rewriting the function and made some additional smaller improvement to the PR. Would be ready to merge from my side 🚀

**Comment by vishnuravi** [comment]
Thank you @PSchmiedmayer for the improvements, and also adding the code coverage report. I have added some unit tests that should improve the coverage. I will go ahead and merge this PR.

**Pull request merged**